src/librustc/traits/query/type_op/outlives.rs

```diff
@@ -30,23 +29,39 @@ impl<'tcx> DropckOutlives<'tcx> {
     }
 }
 
-impl<'gcx, 'tcx> super::TypeOp<'gcx, 'tcx> for DropckOutlives<'tcx> {
-    type Output = Vec<Kind<'tcx>>;
+impl super::QueryTypeOp<'gcx, 'tcx> for DropckOutlives<'tcx>
+where
+    'gcx: 'tcx,
+{
+    type QueryKey = ParamEnvAnd<'tcx, Ty<'tcx>>;
```

Comment:
How come `DropckOutlives` doesn't itself contain a `ParamEnvAnd<'tcx, Ty<'tcx>>`?

Reply:
Well, there is no real reason, it's just mildly annoying to type -- but even if it did, we'd still want the ability to convert "to the query key" before canonicalizing. In particular, if you have a `Canonical<Foo>` where `Foo` has a field `bar: Bar`, you cannot generally convert that to a `Canonical<Bar>`. This is because there may be other fields before or after `Bar` that affect the canonical numbering of regions and things. This is why there is no (for example) `map` operation on `Canonical` -- now, in this case, there is only one field, so in fact it would be fine, but it's just as easy (and more obviously correct) to convert **before** canonicalization.

Reply:
I was thinking that if the type inside `DropckOutlives` is changed, then `QueryKey` would always be equal to `Self` - do you plan for a situation in which that wouldn't be the case?

Reply:
Also, this makes me wonder... `Eq` has a `ParamEnv` but it's not put *around* the types in question - shouldn't it use `ParamEnvAnd<(Ty, Ty)>`? Or am I missing something?